daemon/logger/journald/read.go

```diff
@@ -392,6 +396,76 @@ func (s *journald) readLogs(logWatcher *logger.LogWatcher, config logger.ReadCon
 	return
 }
 
+func (s *journald) readRemoteLogs(logWatcher *logger.LogWatcher, config logger.ReadConfig) {
+	client := http.Client{}
+
+	// Build a url to get logs from systemd-journal-gatewayd.
+	container_id := s.vars["CONTAINER_ID_FULL"]
+	v := url.Values{}
+	v.Set("CONTAINER_ID_FULL", container_id)
+	if config.Follow {
+		v.Set("follow", "")
+	}
+	u := url.URL{
+		Scheme:   "http",
+		Host:     s.gatewayAddress,
+		Path:     "entries",
+		RawQuery: v.Encode(),
+	}
+
+	// Send a request to get logs and get the response from systemd-journal-gatewayd.
+	req, err := http.NewRequest("GET", u.String(), nil)
+	req.Header.Add("Accept", "application/json")
+	resp, err := client.Do(req)
+
+	defer close(logWatcher.Msg)
+	if resp != nil && resp.Body != nil {
+		defer resp.Body.Close()
+	}
+
+	// Handle errors while opening journal.
+	fmtString := "error %q while opening journal from %s for container %q"
+	if err != nil {
+		logWatcher.Err <- fmt.Errorf(fmtString, err, s.gatewayAddress, container_id)
+		return
+	} else if resp != nil && resp.StatusCode >= 400 {
+		logWatcher.Err <- fmt.Errorf(fmtString, resp.Status, s.gatewayAddress, container_id)
+		return
+	}
+
+	// Read logs from the response body.
+	dec := json.NewDecoder(resp.Body)
+	for dec.More() {
+		attrs := make(map[string]string)
+		err := dec.Decode(&attrs)
+		if err != nil {
+			fmtString := "error %q while reading journal from %s for container %q"
+			logrus.Warnf(fmtString, err, s.gatewayAddress, container_id)
+			return
+		}
+		t, err := strconv.ParseInt(attrs["_SOURCE_REALTIME_TIMESTAMP"], 10, 64)
+		timestamp := time.Unix(t/1000000, (t%1000000)*1000)
+		logWatcher.Msg <- &logger.Message{
+			Line:      []byte(attrs["MESSAGE"] + "\n"),
+			Source:    "stdout",
+			Timestamp: timestamp,
+			Attrs:     attrs,
```

Comment:
This appears to encode the entire journal entry as attributes, and every value will be displayed by a client that uses the --details flag to the "docker logs" command.  I'd suggest filtering out some of the fields which aren't added by dockerd (see https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html) here.